internal/xcontext/context_with_timeout.go

```diff
@@ -0,0 +1,88 @@
+package xcontext
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ydb-platform/ydb-go-sdk/v3/internal/xerrors"
+)
+
+func WithTimeout(ctx context.Context, t time.Duration) (context.Context, context.CancelFunc) {
+	childCtx := &timeoutCtx{
+		parentCtx: ctx,
+		from:      xerrors.StackRecord(1),
+	}
+	childCtx.ctx, childCtx.ctxCancel = context.WithTimeout(ctx, t)
+	return childCtx, childCtx.cancel
+}
+
+type timeoutCtx struct {
+	parentCtx context.Context
+	ctx       context.Context
+	ctxCancel context.CancelFunc
+	from      string
+
+	m   sync.Mutex
+	err error
+}
+
+func (ctx *timeoutCtx) Deadline() (deadline time.Time, ok bool) {
+	return ctx.ctx.Deadline()
+}
+
+func (ctx *timeoutCtx) Done() <-chan struct{} {
+	return ctx.ctx.Done()
+}
+
+func (ctx *timeoutCtx) withErrUnderLock(err error) error {
+	switch err { //nolint:errorlint
```

Comment:
Need to difference own cancel and own timeout from parent context cancel and parent timeout